blog/2025/keycloak-bencmark.adoc

```diff
@@ -0,0 +1,198 @@
+====
+
+Memory was increased when we observed high CPU usage by the garbage collector (GC).
+It is difficult to formulate a precise memory requirement because it depends not only on the Keycloak data but also on the number of concurrent requests.
+
+The database instance type was chosen based on observation.
+If we observed failing requests or a 99th percentile response time above 100ms with database CPU usage exceeding 80%, we repeated the test using the next larger database instance type.
+
+A summary of the results can be observed in the table below.
+
+.Keycloak performance with load
+|===
+|OCP Instance Type | DB Instance Type | Pod CPU limit | Pod Memory Limit (GB) | # Pods | Logins/sec | Refreshes Token/sec| Response Time (ms, 99th pct)
+
+m|c8g.8xlarge
+m|db.r8g.2xlarge
+|24
+|4
+|3
+|500
+|2500
+|81
+
+m|c8g.8xlarge
+m|db.r8g.4xlarge
+|40
+|8
+|3
+|1000
+|5000
+|60
+
+m|c8g.24xlarge
+m|db.r8g.16xlarge
+|74
+|8
+|3
+|2000
+|10000
+|58
+
+m|c8g.24xlarge
+m|db.r8g.16xlarge
+|40
+|8
+|6
+|1000
+|20000
+|59
+
+|===
+
+Based on the overall results, we can confirm that Keycloak scales vertically almost linearly.
+
+We used the default Keycloak settings, with the exception of the last two tests.
+For those tests, we had to increase the number of threads (using the `http-pool-max-threads` option) to 330 for the 2,000 logins + 10,000 token refreshes scenario and to 200 for the 1,000 logins + 20,000 token refreshes scenario.
+
+It is also worth noting that we scaled up the OpenShift Ingress Routers because we were observing connection errors on the load generator.
+
+The following images illustrate how the requests correlated with the configured CPU and memory limits.
+
+.CPU limits
+image:${blogImages}/kc_perf_2025/kc_perf_1.png[requests vs cpu]
+
+.Memory limits
+image:${blogImages}/kc_perf_2025/kc_perf_4.png[requests vs memory]
+
+== Latency
+
+Now that we have an understanding of the resources needed, our focus shifts to achieving high availability.
+A key aspect of this is *deploying Keycloak across different availability zones*, which, by its nature, introduces additional latency between Keycloak Pods.
+This leads us to a crucial question: *how does Keycloak behave under these conditions?*
+
+To find out, we used https://chaos-mesh.org/[Chaos Mesh] to introduce an artificial delay to all outgoing network packets.
+This means that each remote call is affected twice: once for the outgoing message and again for the incoming response.
+For instance, a 5ms delay applied to outgoing packets results in a *10ms round-trip time for any remote communication*.
+
+For this test, we used the 500 logins/second and 2,500 token refreshes/second setup.
+The table below summarizes the gathered data, where the 99th percentile of the response time is taken from the Gatling report.
+
+.Network latency
+|===
+| Delay(ms) | Response Time (ms, 99th pct)
+
+|0
+|81
+
+|5
+|101
+
+|10
+|139
+
+|15
+|1135
```

Comment:
The latency spied with 15 ms because the system was then overloaded ... I would rather take a scaled down version here (didn't we test like 400 logins per second), which would show both a increased response time and a reduced throughput of the system. Then it would also look more linear. 

BTW, would we need new number given the latest improvements like the parallel write to the DB and the caches?